### PR TITLE
[WIP] Beacon data

### PR DESCRIPTION
**Pull request by jamierocks**
Implementation of `REPLACE_WITH_DATA_MANIPULATOR`.
- [x] Registration of field getters and setters
- [x] Accurately used the appropriate `AbstractData` implementation

Implementation of `REPLACE_WITH_IMMUTABLE_MANIPULATOR`
- [x] Accurately extended the appropriate `AbstractImmutableData` implementation
- [x] Accurately instantiating final instance fields with an `ImmutableValue` counter part for any value getters
- [ ] If necessary, creating a new `ImmutableValue` for a value that should not be cached, or, using the existing caching utils.

Implementation of the `DATA_MANIPULATOR_BUILDER`
- [ ] In the `build()` method, checked if the container contained the keys required, if it doesn't, return `Optional.empty()`

Implementation of the `VALUE_PROCESSOR`(s):
- [ ] Appropriately extended `AbstractSpongeValueProcessor`
- [ ] Individual processors for each source type possible (one for `ItemStack`/`TileEntity`/`Entity` as necessary).

Registration:
- [ ] Registered the `Key` correctly by `FIELD_NAME` in the `KeyRegistry`
- [ ] Registered the `DataProcessor`s and `DataManipulatorBuilder` in `SpongeSerializationRegistry`
- [ ] Registered the `ValueProcessor`s in the `SpongeSerializationRegistry`

Test Plugin code:

Wrap your plugin code with ```java
// PUT CODE IN HERE

And at the end of your plugin ```

Notes:
Your test plugin should test using all possible value related methods with your `DataManipulator`s and also test the serialization and deserialization of your `DataManipulator`s.

### CODE REVIEW

**Comment by JBYoshi** [comment]
@jamierocks When you finish the PR, could you also update https://github.com/SpongePowered/SpongeCommon/blob/master/src/main/java/org/spongepowered/common/data/builder/block/tileentity/SpongeBeaconBuilder.java#L61?

**Comment by jamierocks** [comment]
@JBYoshi sure :)

File: src/main/java/org/spongepowered/common/data/manipulator/mutable/tileentity/SpongeBeaconData.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+import org.spongepowered.common.data.value.mutable.SpongeOptionalValue;
+import org.spongepowered.common.interfaces.tileentity.IMixinTileEntityBeacon;
+
+import java.util.Optional;
+
+public class SpongeBeaconData extends AbstractData<BeaconData, ImmutableBeaconData> implements BeaconData {
+
+    private final TileEntityBeacon beacon;
+
+    public SpongeBeaconData(TileEntityBeacon beacon) {
```
Comment (Aaron1011) [inline src/main/java/org/spongepowered/common/data/manipulator/mutable/tileentity/SpongeBeaconData.java]: You can't reference a tile entity here- that' what processors are for. The data class should be self-contained.

Comment (jamierocks) [inline src/main/java/org/spongepowered/common/data/manipulator/mutable/tileentity/SpongeBeaconData.java]: Ah okay. Still getting used to things :)

**Comment by gabizou** [comment]
@jamierocks can you provide the things requested from the [issue on what to include in your PR description](https://github.com/SpongePowered/SpongeCommon/issues/8).

**Comment by jamierocks** [comment]
@gabizou Sure, I hadn't noticed them :P

File: src/main/java/org/spongepowered/common/data/builder/manipulator/mutable/tileentity/BeaconDataBuilder.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+                    Optional.of(
+                            (PotionEffectType) Potion.potionTypes[((IMixinTileEntityBeacon) dataHolder).getSecondaryEffect()]);
+            return Optional.of(new SpongeBeaconData(primary, secondary));
+        }
+        return Optional.empty();
+    }
+
+    @Override
+    public Optional<BeaconData> build(DataView container) throws InvalidDataException {
+        DataUtil.checkDataExists(container, Keys.BEACON_PRIMARY_EFFECT.getQuery());
```
Comment (ZephireNZ) [inline src/main/java/org/spongepowered/common/data/builder/manipulator/mutable/tileentity/BeaconDataBuilder.java]: @gabizou Still unsure what the contract of this is - whether to return an empty optional (as used in some data manipulators), or to throw InvalidDataException (as is the case with this and others that use DataUtil).

Comment (jamierocks) [inline src/main/java/org/spongepowered/common/data/builder/manipulator/mutable/tileentity/BeaconDataBuilder.java]: It throws InvalidDataException if the query isn't present in the container.

Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/builder/manipulator/mutable/tileentity/BeaconDataBuilder.java]: @jamierocks I've been discussing this with @Aaron1011 and we've elected that we throw `InvalidDataException`s if the data is improperly formatted, but if the container doesn't contain the data, return `Optional.empty()`.

**Comment by jamierocks** [comment]
I will update this weekend - unless I get chance during the week.

**Comment by gabizou** [comment]
Any updates?

**Comment by jamierocks** [comment]
Yeah - I will resume development shortly.

I may have completely forgotten about this PR, during the last week. No thanks to the amount of gaming I have been doing...

**Comment by gabizou** [comment]
Any updates @jamierocks?

File: src/main/resources/mixins.common.core.json [hunk 3]
```diff
@@ -302,6 +302,7 @@
         "text.MixinClickEvent",
         "text.MixinHoverEvent",
         "util.MixinPacketThreadUtil",
+        "tileentity.MixinTileEntityBeacon",
```
Comment (phroa) [inline src/main/resources/mixins.common.core.json]: that's not alphabetical :|

Comment (jamierocks) [inline src/main/resources/mixins.common.core.json]: bollocks, forgot to check that after merge conflict :/

File: src/main/java/org/spongepowered/common/mixin/core/tileentity/MixinTileEntityBeacon.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+import net.minecraft.tileentity.TileEntityBeacon;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Shadow;
+import org.spongepowered.common.interfaces.tileentity.IMixinTileEntityBeacon;
+
+@Mixin(TileEntityBeacon.class)
+public class MixinTileEntityBeacon implements IMixinTileEntityBeacon {
+
+    @Shadow private int primaryEffect;
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/mixin/core/tileentity/MixinTileEntityBeacon.java]: I'd much rather an AT change than an additional mixin. @SpongePowered/developers, thoughts?

Comment (simon816) [inline src/main/java/org/spongepowered/common/mixin/core/tileentity/MixinTileEntityBeacon.java]: @jamierocks There already is a beacon tile entity mixin, `org.spongepowered.common.mixin.core.block.tiles.MixinTileEntityBeacon`

Comment (jamierocks) [inline src/main/java/org/spongepowered/common/mixin/core/tileentity/MixinTileEntityBeacon.java]: That must be new, don't think it was there when I originally made the mixin.

File: src/main/java/org/spongepowered/common/data/processor/data/tileentity/BeaconDataProcessor.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+import org.spongepowered.api.data.merge.MergeFunction;
+import org.spongepowered.api.data.value.BaseValue;
+import org.spongepowered.api.potion.PotionEffectType;
+import org.spongepowered.common.data.manipulator.mutable.tileentity.SpongeBeaconData;
+import org.spongepowered.common.data.processor.common.AbstractSpongeDataProcessor;
+import org.spongepowered.common.interfaces.tileentity.IMixinTileEntityBeacon;
+
+import java.util.Optional;
+
+public class BeaconDataProcessor extends AbstractSpongeDataProcessor<BeaconData, ImmutableBeaconData> {
```
Comment (simon816) [inline src/main/java/org/spongepowered/common/data/processor/data/tileentity/BeaconDataProcessor.java]: Use `AbstractTileEntityDataProcessor`

**Comment by gabizou** [comment]
Closing because this has waited long enough to be updated. I'll get around to writing it today if anything.

**Comment by jamierocks** [comment]
Eh okay - fair enough.

**Pull request closed without merging**